Protobuf handling

dr7ana opened the pull request.

dr7ana committed:
Protobuf integration & config wrapping/unwrapping

This adds support for automatic wrapping and unwrapping of config
messages that may be wrapped in protobuf; this was unintentional, but is
out in the wild in current Session releases.  The long-term plan is to
stop doing this eventually, but supporting it here allows us to change
from the current mode (always wrap/unwrap affected message types) to
a mode where we only unwrap, and then finally we can drop unwrapping as
well.

For now: we wrap/unwrap the existing released message types (user
profile, conversation info, and so on) but not the new groups configs
(which should always be un-wrapped).

Various notes:

- protobuf added as external dependency.

- protobuf 21.12 is used, which is an older release, but is what is
  still being used by lots of projects and Linux distributions because
  Google applied huge amounts of extra bloat in v22 (including pulling a
  massive pile of crap in abseil), broke the build system in various
  ways, and made everything suck to the point that many Linux
  distributions have just stuck with the latest 21.12 because of all the
  broken crap Google imposed on everyone in v22 and newer.

- SessionProtos.pro and WebSocketResources.proto looped in from
  session-ios codebase, with one minor change to let us use it in "lite"
  mode (which is still very bloated, but nothing like non-lite mode).

- Added pregenerated header/cpp files; we experimented with building
  them as part of the build, but that ended up neigh impossible for our
  various cross-compiled builds.  (These also get excluded from
  format.sh because they break if formatted.  Hurray.)

- Add protobuf::libprotobuf-lite to the static build list.

- Adds a dummy test script that does nothing, but tests that we can
  probably link against the combined archive (when building with
  `-DSTATIC_BUILD=ON`).

jagerman committed:
Disable -Werror for stringop-overflow

It's stupidly difficult to avoid false positives (and protobuf triggers
it).

jagerman committed:
Turn on test timing reporting for ci test suite

jagerman committed:
Add wine to CI deps

We use wine to test the windows CI builds

jagerman commented:
~~TODO: we need to squash these changes together (had to update the protobufs again, for the downgrade).~~

jagerman commented:
Now squashed and rebased on top of #59 (which will most likely be merged before this PR).

jagerman committed:
Small fixes/improvements

- Removed an unneeded `ustring` argument overload which just calls the
  `ustring_view` alternative (which already is fine when given a
  ustring).

- Don't eat exceptions from protobuf production.  If this throws on
  *producing* a protobuf then it is indicative of some application bug
  (unlike input, where we legitimately might not have a protobuf at
  all).

- Switch to protobuf's ParseFromArray to avoid string construction.

- Small code simplifications

jagerman committed:
Add session protocol encryption/decryption

Part of protobuf handling apparently also requires session protocol
decryption, so this adds that to libsession.

jagerman committed:
protobuf, v2

Adds protobuf parsing as Session does it.

jagerman committed:
Fix missing libsodium link for protobuf code

jagerman committed:
Add deterministic session-encryption; use it for config protobuf

This adds a deterministic version of session protocol encryption that
differs from crypto_box_seal by using a keyed hash of the sender seed,
recipient pubkey, and message to generate the ephemeral key.

This effectively neuters such an encrypted message to be a regular
crypto_box encryption, which means the sender is later able to decrypt
it; this is thus only really a special case for protobuf+encrypted
config messages for backwards compatible config message support.

jagerman committed:
Move session/protos.hpp -> session/config/protos.hpp

The only thing in this header are config-specific, so seems better under
`config`.  This also resolves some issues with circular library
dependencies in the build between config, crypto, and protos.

jagerman committed:
Allow using system protobuf for shared lib builds

Also moves the `protos` build into /protos rather than building it via
`../` inside src.

jagerman committed:
Make the onionreq code optional; disable for ios.sh

The dependencies are causing headaches right now on iOS, and currently
session clients don't need this code so can disable it.

jagerman committed:
Also disable gmp/nettle in StaticBuild

jagerman committed:
Set timestamp to 1 to pass old Session clients validity check

jagerman committed:
Return good hashes from merge()

Session clients required the list of hashes from a merge() to perform
various operations based on the messages that were accepted.  This
updates `merge()` to return a vector of accepted hashes.

For the C API, this returns a config_string_list* pointer that needs to
be freed, but contains all the good hashes (the same way
current_hashes() works).

jagerman committed:
Fix build host regex to recognize "ios13.0"

It was only matching 13 but not 13.0, which was making gmp putting in
assembly instructions that ios doesn't support (because gmp only turns
that off if it sees "darwin" in the host string).

jagerman committed:
Add macos-compatible nproc alternative

Without this we were running `make -j` on macos which is a minor fork
bomb.  Also adds a workaround in case that doesn't work either to fall
back to just -j1.
